### ISSUE TITLE
Rule proposal: Check classNames

### ISSUE THREAD

**Issue by remcohaszing** [yannickcr/eslint-plugin-react#2421]
I think tt would be nice if there is a rule that would check the `className` prop. It could check 2 things:

1. Class name validity.
2. Useless class name spacing.

It would also be nice if it could support the [classnames](https://www.npmjs.com/package/classnames) package.

The following patterns are considered warnings:

```jsx
<div className="" />
```

```jsx
<div className=" " />
```

```jsx
<div className=" foo" />
```

```jsx
<div className="bar " />
```

```jsx
<div className="foo  bar" />
```

```jsx
<div className="1" />
```

```jsx
<div className="?" />
```

```jsx
classNames({
  'foo ': foo,
  ' bar': bar,
})
```

The following patterns are not considered warnings:

```jsx
<div className="foo" />
```

```jsx
<div className="bar" />
```

```jsx
<div className="foo bar" />
```

```jsx
<div className={`foo ${bar}`} />
```

```jsx
<div className={`foo ${bar || ''}`} />
```

```jsx
<div className={`foo ${bar ? 'bar' : ''}`} />
```

```jsx
classNames({
  'foo': foo,
  bar: bar,
  [baz]: bazz
})
```

There could be an option to determine whether or not the `className` prop is also checked on custom components. This would toggle validity on the following situation:

```jsx
<MyComponent className=" bar" />
```

**Comment by ljharb** [comment]
What's invalid about that, exactly?

**Comment by remcohaszing** [comment]
A single class name should match the following regex (from https://stackoverflow.com/a/449000/1154610):

```js
/-?[_a-zA-Z]+[_a-zA-Z0-9-]*/
```

So these class names are invalid:

- `?`
- `1`
- `1a`

Class names are separated by a space, `foo bar·` is equivalent to `····foo·······bar······`, where `·` represents a space.

**Comment by ljharb** [comment]
right, but all the extra whitespace isn't invalid - only leading non-alpha, or non-alphanumeric characters.

Given that `className` is often used very dynamically - and that the `classname` lib is the most common pattern with it - i'm not sure how much value a lint rule would offer.

**Comment by kyleknighted** [comment]
It would also help with interfaces that have multiple developers outside of the team (open source plugin contributions) that may use a deprecated global class in their code that should raise a warning/error so that they can use the appropriate component(s).

**Comment by ljharb** [comment]
This seems like something react itself needs to check; I don't think static analysis will be very helpful here.
